File: java/Interpreter.java
import java.util.*;
import java.io.*;

public class Interpreter
{
	private String program;
	
	private int pc;
	private int programLength;
	
	private ArrayList<Character> memory;
	private int pointerPosition;
	
	private Deque<Integer> loopStack;
	private int[] loopArray;
	
	public Interpreter(String pg)
	{
		program = pg;
		pc = 0;
		programLength = program.length();
		pointerPosition = 0;
		
		memory = new ArrayList<Character>(); 
		loopStack = new ArrayDeque<Integer>();
		loopArray = makeLoopArray();
	}
	
	private int[] makeLoopArray()
	{
		int[] arr = new int[programLength];
		Deque<Integer> loopStartIndex = new ArrayDeque<Integer>();
		for(int i = 0; i < programLength; i++)
		{
			char c = program.charAt(i);
			if(c == '[')
			{
				loopStartIndex.addFirst(i);
				arr[i] = 0;
			}
			else if(c == ']')
			{
				int index = loopStartIndex.removeFirst();
				arr[index] = i;
				arr[i] = index;
			}
			else
			{
				arr[i] = 0;
			}
		}
		
		return arr;
	}
	
	private int interpret(char inst)
	{
		int error = 0;
		switch(inst)
		{
			case '<':
				if(pointerPosition > 0) 
					pointerPosition -= 1;
				//else
				//	pointerPosition = 0;
				break;
			case '>':
				if(pointerPosition < memory.size() - 1) 
					pointerPosition += 1;
				else// if(pointerPosition >= memory.size() - 1)
				{
					memory.add((char)0);
					pointerPosition += 1;
				}
				break;
			case '+':
				if(pointerPosition >= memory.size())
					memory.add((char)(1));
				else
					memory.set(pointerPosition, (char)(memory.get(pointerPosition) + 1));
				break;
			case '-':
				if(pointerPosition >= memory.size())
					memory.add((char)(-1));
				else
					memory.set(pointerPosition, (char)(memory.get(pointerPosition) - 1));
				break;
			case '[':
				if(memory.get(pointerPosition).equals((char)0))//go to end of loop
					pc = loopArray[pc];
				//else
				//	loopStack.addFirst(pc);
				break;
			case ']':
				if(!memory.get(pointerPosition).equals((char)0))//go back to start of loop
					pc = loopArray[pc];
				//else
				//	loopStack.removeFirst();
				break;
			case '.':
				//output
				if(pointerPosition < memory.size())
					System.out.print(memory.get(pointerPosition));
				break;
			case ',':
				//input
				Scanner sc = new Scanner(System.in);
				char b = (char)sc.nextByte();
				memory.set(pointerPosition, b);
				sc.close();
				break;
			default:
			//noop
				break;
		}
	
		pc++;
		return error;
	}
	
	public int interpret()
	{
		int error = 0;
		while(pc < programLength)
			error = interpret(program.charAt(pc));
		return error;
	}
}

File: c/Interpreter.c
#include "Interpreter.h"

char* memory = NULL;
int* looparray = NULL;

int memsize = 0;
int mempos = 0;
int pc = 0;

int makeLoopArray(char* program, int programSize)
{
  looparray = (int*)malloc(programSize * sizeof(int));
  memset(looparray, -1, programSize * sizeof(int));
  //max amount of start loop operands must be half of the program size since if there were more we wouldnt have enough matching end loop
  int* loopstack = (int*)malloc((programSize * sizeof(int) / 2) + 1);
  memset(loopstack, 0, (programSize * sizeof(int) / 2) + 1);
  int top = 0;//points to the empty spot above the stack
  int i = 0;
  for(i = 0; i < programSize; i++)
  {
    char c = program[i];
    if(c == '[')
    {
      loopstack[top++] = i;
      looparray[i] = 0;
    }
    else if(c == ']')
    {
      if(top == 0) return -1;
      int index = loopstack[--top];
      looparray[index] = i;
      looparray[i] = index;
    }
  }
  free(loopstack);
  if(top != 0) return -1;
  return 0;
}

int interpret(char* program, int programSize)
{
  //initialize memory with 30000 spots
  memsize = 30000;
  memory = (char*)malloc(memsize);
  memset(memory, 0, memsize);
  //make loop array that keeps track of matching brackets
  if(makeLoopArray(program, programSize) == -1)
  {
    printf("ERROR: Unmatched loop brackets\n");
    return -1;
  }
  //start interpreting
  for(pc = 0; pc < programSize; pc++)
  {
    char inst = program[pc];
    char c;
    switch(inst)
    {
      case '<':
        if(mempos > 0) mempos--;
        break;
      case '>':
        if(mempos < memsize - 1)
          mempos++;
        else
        {
          //allocate more memory
          memsize *= 2;
          memory = realloc(memory, memsize);
          if(memory == NULL) return -1;
          mempos++;
        }
        break;
      case '+':
        c = memory[mempos];
        c += 1;
        memory[mempos] = c;
        break;
      case '-':
        c = memory[mempos];
        c -= 1;
        memory[mempos] = c;
        break;
      case '[':
        if(memory[mempos] == 0) pc = looparray[pc];
        break;
      case ']':
        if(memory[mempos] != 0) pc = looparray[pc];
        break;
      case '.':
        putchar(memory[mempos]);
        break;
      case ',':
        memory[mempos] = getchar();
        break;
      default:
        break;
    }
  }
  return 0;
}
